File: public/code/smc__inducement.py
"""
Inducement & Mitigation Strategy
=================================
Inducement = düşük zaman diliminde sahte break (likidite çekmek için).
Mitigation = sahte break sonrası gerçek yönde hareket başlar.

Setup:
  1. Bir swing low/high oluşur
  2. Sahte break (wick ile aşar ama kapatamaz)
  3. Hızlı geri dönüş = inducement
  4. Önceki yapı kırılır = mitigation onayı -> entry
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, atr, swing_highs_lows


def detect_inducement_signal(df: pd.DataFrame, swing_lb: int = 8) -> dict:
    if len(df) < swing_lb * 3:
        return make_signal(reason="Insufficient data")

    sh, sl = swing_highs_lows(df, swing_lb)
    current_atr = atr(df, 14).iloc[-1]
    current = df.iloc[-1]
    last_3 = df.iloc[-4:-1]  # son 3 mum (current hariç)

    recent_lows = sl.dropna().tail(3)
    recent_highs = sh.dropna().tail(3)
    if len(recent_lows) < 2 or len(recent_highs) < 2:
        return make_signal(reason="No structure")

    last_low = recent_lows.iloc[-1]
    last_high = recent_highs.iloc[-1]

    # Bullish inducement: son 3 mumda bir tanesi low'un altına wick attı ama kapatamadı
    wick_below = ((last_3["low"] < last_low) & (last_3["close"] > last_low)).any()
    rejected = current["close"] > current["open"] and current["close"] > last_3["high"].max()
    if wick_below and rejected:
        entry = current["close"]
        stop_loss = last_3["low"].min() - 0.3 * current_atr
        risk = entry - stop_loss
        return make_signal(
            signal="long",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry + risk * 2, entry + risk * 3, entry + risk * 5],
            confidence=0.7,
            reason="Bullish inducement + mitigation",
        )

    wick_above = ((last_3["high"] > last_high) & (last_3["close"] < last_high)).any()
    rejected_dn = current["close"] < current["open"] and current["close"] < last_3["low"].min()
    if wick_above and rejected_dn:
        entry = current["close"]
        stop_loss = last_3["high"].max() + 0.3 * current_atr
        risk = stop_loss - entry
        return make_signal(
            signal="short",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry - risk * 2, entry - risk * 3, entry - risk * 5],
            confidence=0.7,
            reason="Bearish inducement + mitigation",
        )

    return make_signal(reason="No inducement setup")

File: public/code/smc__inducement.ts
/**
 * Inducement & Mitigation Strategy
 */
import { Candle, Signal, makeSignal, atr, swingHighs, swingLows } from "../../common";

export function detectInducementSignal(candles: Candle[], swingLb = 8): Signal {
  if (candles.length < swingLb * 3) return makeSignal({ reason: "Insufficient data" });

  const sh = swingHighs(candles, swingLb).filter((v): v is number => v !== null);
  const sl = swingLows(candles, swingLb).filter((v): v is number => v !== null);
  if (sh.length < 2 || sl.length < 2) return makeSignal({ reason: "No structure" });

  const atrValues = atr(candles, 14);
  const currentAtr = atrValues[atrValues.length - 1];
  const current = candles[candles.length - 1];
  const last3 = candles.slice(-4, -1);
  const lastLow = sl[sl.length - 1];
  const lastHigh = sh[sh.length - 1];

  const wickBelow = last3.some((c) => c.low < lastLow && c.close > lastLow);
  const max3High = Math.max(...last3.map((c) => c.high));
  const rejected = current.close > current.open && current.close > max3High;
  if (wickBelow && rejected) {
    const min3Low = Math.min(...last3.map((c) => c.low));
    const entry = current.close;
    const stopLoss = min3Low - 0.3 * currentAtr;
    const risk = entry - stopLoss;
    return makeSignal({
      signal: "long", entry, stop_loss: stopLoss,
      take_profit: [entry + risk * 2, entry + risk * 3, entry + risk * 5],
      confidence: 0.7, reason: "Bullish inducement + mitigation",
      timestamp: current.timestamp,
    });
  }

  const wickAbove = last3.some((c) => c.high > lastHigh && c.close < lastHigh);
  const min3Low = Math.min(...last3.map((c) => c.low));
  const rejectedDn = current.close < current.open && current.close < min3Low;
  if (wickAbove && rejectedDn) {
    const max3HighB = Math.max(...last3.map((c) => c.high));
    const entry = current.close;
    const stopLoss = max3HighB + 0.3 * currentAtr;
    const risk = stopLoss - entry;
    return makeSignal({
      signal: "short", entry, stop_loss: stopLoss,
      take_profit: [entry - risk * 2, entry - risk * 3, entry - risk * 5],
      confidence: 0.7, reason: "Bearish inducement + mitigation",
      timestamp: current.timestamp,
    });
  }

  return makeSignal({ reason: "No inducement setup" });
}
